### DISCUSSION TITLE
ModBus info

### DISCUSSION THREAD

**Discussion by mackyb27** [nanoframework/Home#1442] [Q&A]
Hi folks!

Hi have a few question regarding current Modbus implementation for nanoframework, if there is someone that can help me. First is - is there a way to dynamically define a device ID on the network (all nanoframework devices like esp32, a master and a few slaves)? Using a scan function, maybe using MAC address in the process or something similar? I'd sure would like to have so more elaborate examples in Modbus folder. Second - I have trouble understanding how to implement a ModBus Device, since modbus client have a function ReadHoldingRegisters that returns short[], but the Device implementation can have an overload function TryReadHoldingRegisters that has only a short as an out parameter, not short[]. I would like to return a sensor info which is a double. Is there a way of how to do that? I guess the way would be to convert double to short[] but there is no function to overload in this regard... maybe I'm just not understanding how it suppose to work.

kind regards, martin

**Comment by josesimoes** [comment]
Our current implementation offers API to communicate with Modbus clients. Nothing on server.

On the second part (about conversion from double value to byte arrays) this has nothing to do with C#. You should look into Modbus specs and learn how that's supposed to be implemented. Sensors outputting float values are absolutely common, so this has to be there.

**Reply by mackyb27** [reply]
Thx for quick reply. There is an implementation of [Modbus Server](https://github.com/nanoframework/nanoFramework.IoT.Device/blob/develop/devices/Modbus) for nanoframework where it features a ModbusDevice implementation. I was asking about this specific implementation. There is an example attached but unfortunately too sparse. I have read the modbus specs, I'm not sure about the implementation of ModbusDevice that is needed in order to run ModbusServer. Are then client and server compatible at all when there are value types other than boolean or short?

**Reply by josesimoes** [reply]
You are correct! That one does implement both the server and the client. My bad. 🤦

Never used the server part. I don't think you need the client to implement a server... a single device can act as one or the other, not both.
Unless you're connecting it to different COM ports, that is.

Data is transmitted in binary packets, it's up to the layer above to make sense of it.

**Reply by mackyb27** [reply]
Ok. I figured out. But it would damn be nice to include it in the samples. So this is for those who are looking to implement the modbus while waiting for CAN implementation on ESP32 ;). Samples only deal with one byte implementation, but usually sensor readings are of float type. Also MAC address can be transmitted similar way, however it requires special care. If the data is larger than a byte, modbus protocol has a start address and a number of bytes to retrieve. Examples provided do not cover this part so I try to do it here. Basically modbus terminology is using function codes to determine the function to perform on the server (slave) and a request object that encapsulate the parameters and the response object which decodes the message.

on the client side put this into sendmodbusrequest or similar:
`case ModbusFunctionCode.ReadHoldingRegisters:
    var result = client.ReadHoldingRegisters(request.SlaveAddress, request.StartAddress, request.Quantity);
    response = new ModbusResponse(request.SlaveAddress, ModbusFunctionCode.ReadHoldingRegisters, result);
    break;`

SlaveAddress is the address of modbus device. Generally it is predefined, thou there might be some helper tools to somewhat configure devices.
StartAddress represents the initial address on the server where the first byte of the response will be provided.
Quantity represents the number of bytes t return. For example for a double you'd require 4 bytes 

When implementing server, you have to implement your own Modbus device. You will have to override the function TryReadHoldingRegister (ushort address, out short value) or TryReadInputRegister, depending on the sensor type. However in this case doesnt really matter, choose one or the either. I've split the ushort address into 100 length blocks (100 bytes), however you might do it differently. For this example you'd need only 4.

So corresponding override should look something like:
`protected override bool TryReadHoldingRegister(ushort address, out short value)
{     
    
    switch(address)
    {
        
        case 200://sensor readout
            double sensorReadout1 = ReadSensor();                   
            value = EncodeDouble(sensorReadout1)[0];
            break;
        case 201://sensor readout
            double sensorReadout2 = ReadSensor();   
            value = EncodeDouble(sensorReadout2)[1];
            break;
        case 202://sensor readout
            double sensorReadout3 = ReadSensor();   
            value = EncodeDouble(sensorReadout3)[2];
            break;
        case 203://sensor readout
            double sensorReadout4 = ReadSensor();   
            value = EncodeDouble(sensorReadout4)[3];
            break;
        default:
            value = 0;
            return false;
            
    }

    return true;
}`

and the conversion method to and from short[]:
`public static class DoubleUtils
{
    private const int DoubleBytes = 8;
    private const int DoubleShorts = DoubleBytes / sizeof(ushort);

    public static ushort[] EncodeDouble(double value)
    {
        byte[] bytes = BitConverter.GetBytes(value);
        ushort[] result = new ushort[DoubleShorts];

        for (int i = 0; i < DoubleShorts; i++)
        {
            result[i] = (ushort)(bytes[i * 2] << 8 | bytes[i * 2 + 1]);
        }

        return result;
    }

    public static double DecodeDouble(ushort[] encodedDouble)
    {
        if (encodedDouble == null || encodedDouble.Length != DoubleShorts)
        {
            throw new ArgumentException("Invalid encoded double format.");
        }

        byte[] bytes = new byte[DoubleBytes];

        for (int i = 0; i < DoubleShorts; i++)
        {
            bytes[i * 2] = (byte)(encodedDouble[i] >> 8);
            bytes[i * 2 + 1] = (byte)(encodedDouble[i] & 0xFF);
        }

        return BitConverter.ToDouble(bytes);
    }
}`


hope this helps. cheers, Martin

**Reply by networkfusion** [reply]
Anything holding you back from making a PR for the samples?

**Reply by josesimoes** [reply]
@mackyb27 glad that you've found your way through this.
As networkfusion already mentioned: please go for a PR improving the samples. This is a community effort and everyone help is much welcomed. 😉
